### ISSUE TITLE
Dockerfile

### ISSUE THREAD

**Issue by lassik** [koka-lang/koka#96]
Here's a Dockerfile that clones the Git master and boots into the REPL.

Beware: it builds GHC from source, and the resulting container is 5.74GB! If a Haskell Stack adept found the right incantations to clean up intermediate files, it could probably be cut down to size.

```Dockerfile
FROM debian:sid-slim
RUN apt-get update \
 && apt-get install -y --no-install-recommends \
      ca-certificates curl git gzip xz-utils \
      make gcc libc-dev libgmp-dev libtinfo-dev \
      cmake ninja-build nodejs \
 && rm -rf /var/lib/apt/lists/*
WORKDIR /build
RUN curl --location --fail --silent --show-error -o stack.tar.gz \
      https://get.haskellstack.org/stable/linux-x86_64.tar.gz
RUN mkdir stack \
 && tar --strip-components 1 -C stack -xf stack.tar.gz \
 && mv stack/stack /usr/local/bin/stack \
 && rm -rf stack stack.tar.gz
RUN stack --compiler ghc-8.6.5 setup
RUN git clone --recursive https://github.com/koka-lang/koka
WORKDIR /build/koka
RUN stack build
RUN stack install
RUN stack clean
CMD ["stack", "exec", "koka"]
```

**Comment by lassik** [comment]
Debian `sid` is needed instead of the stable `buster` since the latter does not have a recent enough version of CMake. If we used `buster` we'd have to build CMake from source in addition to GHC.

**Comment by daanx** [comment]
That's great! Thank so much.  5.74GiB lol. One day we will build Koka in Koka ...
What CMake version is Debian `buster` coming with?  I just changes the koka build to not use CMake anymore but invoke the C compiler directly -- the only point where CMake is used is now to build `kklib` once so we can probably clean up `kklib/CMakeLists.txt` to only depend on a lower version of CMake.

**Comment by lassik** [comment]
Debian `buster`:

```
$ cmake --version
cmake version 3.13.4
```

Debian `stretch`:

```
$ cmake --version
cmake version 3.7.2
```

The official `haskell:8.6.5` container is based on Debian stretch, and is "only" 1.26GB. If you're fine with the stretch version of cmake, we could base the Koka container on that official Haskell container.

**Comment by lassik** [comment]
If Koka doesn't need Haskell at runtime, we could find some way to use it only in the build container, then throw it away and keep only Koka.

What' the full set of things that Koka needs at runtime? GCC, Node.js, ...?

**Comment by daanx** [comment]
Thanks -- I think we can make the CMake dependency 3.7.2 ... not entirely sure but should be no issue. It would be great to use the smaller container. At runtime Koka needs:

- when compiling to C: well, it generates an executable so I gueses just `libc` (or whatever the C compiler used) and `libpthreads` and `libm` (math).
- when compiling to JS: just nodeJS to run it (or a browser) but it uses the `amdefine.js`  javascript library for the modules.

**Comment by lassik** [comment]
GHC probably takes more space than GCC and Node.js, especially when compiled from source and we keep the intermediate files (I don't know much about Stack, so I don't know how to remove them.)

But if Stack has a command to install Koka into `/usr/local` after it has been built, so that the build directory is not needed anymore, then I can easily make a production container that copies only `/usr/local` from the build container and throws away the rest (including Stack and GHC).

If you push a branch with CMake downgraded to 3.7.2, I can try building a container from that branch to find out if it works.

**Comment by daanx** [comment]
Currently, once you build Koka, you can run `> stack exec koka -- util/install` and that will install to `~/.local`  (since stack does not have enough privilege to install to `/usr/local`). However, once that is done one can use for example `>sudo cp -r ~/.local/* /usr/local`.   

I think that is enough to have `koka` work standalone (surely for C compilation, but not entirely sure if that copies `amdefine` for javascript as well).  That would need no more `CMake` or `GHC`. 

Actually, if you go into `kklib/CMakeLists` and remove the `cmake_policy` line and lower the `cmake_minimum_required(VERSION 3.8)` to `3.7` this may already work?  Can you try?

**Comment by daanx** [comment]
ah, I spoke too soon :-)  Actually, when you use a local install, `kklib` is still in source form and compiled on-the-fly for your specific platform/compiler -- using `cmake`...  So the final system still needs a C compiler and CMake.  Still, it would be good to make the CMake requirement as low as possible, like 3.7.2 so we can run on as many systems _as is_

**Comment by lassik** [comment]
> Currently, once you build Koka, you can run `> stack exec koka -- util/install` and that will install to `~/.local` (since stack does not have enough privilege to install to `/usr/local`). However, once that is done one can use for example `>sudo cp -r ~/.local/* /usr/local`.

The Debian Docker container starts of as root, so it has full privileges to change any files. If Stack has an option to install directly into `/usr/local`, we can use it.

> I think that is enough to have `koka` work standalone (surely for C compilation, but not entirely sure if that copies `amdefine` for javascript as well). (and neither `CMake` or `GHC` or needed anymore)

👍 

> Actually, if you go into `kklib/CMakeLists` and remove the `cmake_policy` line and lower the `cmake_minimum_required(VERSION 3.8)` to `3.7` this may already work? Can you try?

OK, I'll try.

**Comment by daanx** [comment]
It is possible to say `> stack exec koka -- util/install -- --prefix=/usr/local`  (note the two `--`) but not sure if it will work correctly with the right file attributes; an explicit copy afterwards may work better.

**Comment by lassik** [comment]
CMake 3.7 does not seem to work. When starting the Koka REPL for the first time:

```
cmake  : kklib from: /build/koka/kklib
-- The C compiler identification is GNU 6.3.0
-- Check for working C compiler: /usr/bin/gcc
-- Check for working C compiler: /usr/bin/gcc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
CMake Error at CMakeLists.txt:3 (cmake_policy):
  Policy "CMP0077" is not known to this version of CMake.
```

**Comment by daanx** [comment]
I just pushed a possible "fix" to the `dev` branch. Can you try it?  It just removes the `cmake_policy` line and uses version 3.7.. fingers crossed :-)

**Comment by lassik** [comment]
> I just pushed a possible "fix" to the `dev` branch. Can you try it? It just removes the `cmake_policy` line and uses version 3.7.. fingers crossed :-)

Seems to work fine!

**Comment by daanx** [comment]
ah, sweet. That means that we can use `stretch`; and then for the final build remove the intermediates and just keep `usr/local`, `cmake` and the C compiler (`gcc`?). ?   Pretty sure that will do for C compilation but we would need to test with JavaScript.

Note that `> stack exec koka -- util/install` installs `<prefix>/bin/koka` and also `<prefix>/lib/koka/v2.x.x/kklib` and `<prefix>/lib/koka/v2.x.x/lib` folders.

**Comment by lassik** [comment]
I'm having no luck with `stack exec koka -- util/install -- --prefix=/usr/local`: it doesn't install anything into `/usr/local`.

`stack exec koka -- util/install` without the prefix would appear to install only the `koka` executable into `~/.local`, not any of the libraries it needs.

Can you write an install script that copies everything necessary to run Koka under the prefix? I could then install these from apt:

* cmake
* gcc
* libc-dev
* nodejs

Are Ninja, `make`, or the GMP numeric library needed at runtime?

**Comment by daanx** [comment]
That is strange. The `util/install` should be the script to copy everything. When I run this from latest `dev` branch on Ubuntu 20.04 I see:
```
~/dev/koka$ stack build --fast
...
~/dev/koka$ stack exec koka -- util/install
...
linking: util_install
created: out/v2.0.4/gcc-debug/util_install

installing koka:
 bin: /home/daan/.local/bin
 lib: /home/daan/.local/lib/koka/v2.0.4
copy koka executable...
copy standard libraries...
copy kklib runtime...
copy contrib...
done installing.
~/dev/koka$ ls ~/.local/bin/
koka
~/dev/koka$ ls ~/.local/lib/koka/v2.0.4/
contrib  kklib  lib
```
Can you check if you see the same? If not, what is the exact system you are using? If it does not work I should fix the koka runtime primitives for file copying I guess. I can also create a small shell script of course but it would be nice if we can be maximally portable and use just koka itself for these tasks :-)

**Comment by lassik** [comment]
OK, I used the above commands and now it works for me, even with `--prefix=/usr/local`. Must have misplaced some command in the prior file.

The new container is 322MB, much more palatable.

The runtime container starts from a standard Debian install, adds Koka to /usr/local, and installs the following apt packages:

* `ca-certificates`
* `cmake`
* `gcc`
* `libc-dev`
* `libgmp-dev`
* `libtinfo-dev`
* `make`
* `ninja-build`
* `nodejs`

Is there anything extraneous or missing in that list?

Is there a way to pre-compile the following stuff which is done at the start of the REPL:

```
compile: /usr/local/lib/koka/v2.0.4/lib/std/core.kk
compile: /usr/local/lib/koka/v2.0.4/lib/std/core/types.kk
check  : std/core/types
compile: /usr/local/lib/koka/v2.0.4/lib/std/core/hnd.kk
check  : std/core/hnd
check  : std/core

> 1 + 2 + 3;
check  : interactive
check  : interactive
cmake  : kklib from: /usr/local/lib/koka/v2.0.4/kklib
-- The C compiler identification is GNU 6.3.0
-- Check for working C compiler: /usr/bin/gcc
-- Check for working C compiler: /usr/bin/gcc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Configuring done
-- Generating done
-- Build files have been written to: /out/v2.0.4/gcc-debug/kklib
[13/13] Linking C static library libkklib.a
```

**Comment by lassik** [comment]
Also, can you merge `dev` into `master`? :) The CMake changes appear to be working fine.

**Comment by daanx** [comment]
Just merged master to 2.0.4 with the new CMake.

**Comment by daanx** [comment]
> The new container is 322MB, much more palatable.

Great!
 > Is there anything extraneous or missing in that list?

I think it is good!
 
> Is there a way to pre-compile the following stuff which is done at the start of the REPL:
> ```
> cmake  : kklib from: /usr/local/lib/koka/v2.0.4/kklib
> -- The C compiler identification is GNU 6.3.0
> -- Check for working C compiler: /usr/bin/gcc
> -- Check for working C compiler: /usr/bin/gcc -- works
> -- Detecting C compiler ABI info
> -- Detecting C compiler ABI info - done
> -- Detecting C compile features
> -- Detecting C compile features - done
> -- Configuring done
> -- Generating done
> -- Build files have been written to: /out/v2.0.4/gcc-debug/kklib
> [13/13] Linking C static library libkklib.a
> ```

ah yes. It rebuilds the runtime library from scratch. We can surely prevent this by building a debug/release/drelease during installation. I will look into it.  It would be even better to also build `std/core/types` ,`std/core/hnd` and `std/core` during install as well but we may do that later as I would like to do that properly as part of having "packages"

**Comment by daanx** [comment]
(btw. Koka does not use `libgmp` (but has its own implementation for large integer arithmetic))

**Comment by lassik** [comment]
> (btw. Koka does not use `libgmp` (but has its own implementation for large integer arithmetic))

OK, I'll drop libgmp. It was originally needed to build GHC I think.

> what would be the easiest way to deploy this? I.e. can you write install/deploy instructions for others -- this would make it super easy for many to try out Koka right?

The easiest solution would be that you register a Docker Hub account for yourself (name doesn't matter), then create a `koka-lang` organization and add a `koka` container under it. Then people can type this at any computer that has Docker installed:

`docker run -it koka-lang/koka`

and they will get a REPL.

Docker Hub can (and should) be set to build the container from a GitHub repo. It installs a hook into GitHub so that whenever a commit is made to that GH repo, the container will be rebuilt. The repo needs to contain the `Dockerfile` for the container. It's also possible to create multiple variants (Docker calls them "tags", they have nothing to do with Git tags) of the same container from one Git repo.

**Comment by lassik** [comment]
> Once `nodejs` is installed, I think we need to install `amdefine` and `requirejs`.

OK, Debian has APT packages `node-amdefine` and `node-requirejs`. https://packages.debian.org/

**Comment by daanx** [comment]
Ah, that sounds great. I have no experience with this but I will start setting up an account. Is it possible to retain an specific docker container per release or will it always track the latest commit?

**Comment by lassik** [comment]
Great, thanks! It is possible to retain releases (no pun intended :) In Docker terminology, each release would have its own "tag".

If you trust the security of Docker Hub (which may or may not be wise; they had one major breach where account information was leaked) you can connect the `koka-lang/koka` container straight to this `koka-lang/koka` GitHub repo. Then you can always have a container of the latest git master, in addition to whatever release containers you choose to set up.

If I'm not mistaken, Docker Hub needs _owner_ access to a GitHub organization in order to auto-build containers on commit. There's some technicality that makes things not work otherwise. Owner access is pretty heavy. However, you can also put the Dockerfiles in another GitHub org made just for the purpose.

**Comment by daanx** [comment]
Pfff, I tried to set this up but it is not working -- I guess it has to do with the permissions or something.
I need to read up more on how this works but I guess there must be a way to invoke Docker Hub explicitly to start building?

**Comment by lassik** [comment]
Yes, there is always the option of rebuilding manually. Go to the container's home page and then Manage repository -> Builds, and click Trigger next to the build you want. The Configure Automated Builds button on that page lets you tweak the build settings (the same settings are used by Trigger).

If a build failed, you should see it in Recent Builds and click on it to get a build log to investigate. But I'm not sure this happens if it can't access GitHub at all.

What's the point where you got stuck? https://hub.docker.com/u/koka-lang doesn't show an organization yet.

**Comment by lassik** [comment]
Here is the latest `Dockerfile`:

```Dockerfile
FROM haskell:8.6.5 AS build
RUN mkdir -p ~/.local/bin \
 && cp /usr/local/bin/stack ~/.local/bin/stack \
 && find /usr/local -type f -delete
RUN apt-get update \
 && apt-get install -y --no-install-recommends cmake \
 && rm -rf /var/lib/apt/lists/*
WORKDIR /build
RUN git clone --recursive https://github.com/koka-lang/koka
WORKDIR /build/koka
RUN stack build
RUN stack exec koka -- util/install -- --prefix=/usr/local

FROM debian:stretch
RUN apt-get update \
 && apt-get install -y --no-install-recommends \
      gcc libc-dev \
      cmake make ninja-build \
      nodejs ca-certificates \
 && rm -rf /var/lib/apt/lists/*
COPY --from=build /usr/local/ /usr/local/
CMD ["koka"]
```

**Comment by daanx** [comment]
I build a local docker container with your file -- but when I run the docker image it will start `koka` but then just keep outputting red "> " characters infinitely and I need to manually terminate the image. Is this a Windows thing? something with console input?

**Comment by daanx** [comment]
Ah, when I run as `docker run -it <tag>` it works!  I see now that it will always re-compile `std/core/types`, `std/core/hnd` and `std/core`, can we get around that? for example by keeping the `out/v2.x.x/gcc-<variant>` folders that were generated on installation?

**Comment by daanx** [comment]
In the latest `dev`, the dockerfile should build an image that does not need to recompile :-)  and also includes the `samples` folder. It is a bit crude by copying the samples to `/samples` but I guess ok for the docker REPL to play with?

**Comment by lassik** [comment]
Glad to hear you got it to work!

> I build a local docker container with your file -- but when I run the docker image it will start `koka` but then just keep outputting red "> " characters infinitely and I need to manually terminate the image. Is this a Windows thing? something with console input?

> Ah, when I run as `docker run -it <tag>` it works!

Yes, `docker run -it` wires stdin/stdout/stderr into a terminal (pty). Without `-it` stdin does not look like a terminal (at least not on the container's side). When running Koka with `docker run -it`, pressing Ctrl+D in the REPL (which causes end-of-file on stdin) makes Koka print another prompt `>`. In most REPLs, pressing Ctrl+D exits the REPL. The case of repeating `>` promps without `-it` is caused by this. You should probably make the REPL exit on EOF; most REPLs do that.

> The org on Docker Hub is "kokalang" as Docker does not allow "-" somehow.

Ah, that's too bad. I didn't know about the limitation.

> Would it be better have a default user -- that way we can also install all the koka samples (from the `samples` folder).

> I see now that it will always re-compile `std/core/types`, `std/core/hnd` and `std/core`, can we get around that? for example by keeping the `out/v2.x.x/gcc-<variant>` folders that were generated on installation?

> In the latest `dev`, the dockerfile should build an image that does not need to recompile :-) and also includes the `samples` folder. It is a bit crude by copying the samples to `/samples` but I guess ok for the docker REPL to play with?

The container does have `useradd`/`adduser` so we could add a non-root user, but it's not usually necessary since the container is isolated from the host OS anyway.

Using `/out` or `/samples` in the container doesn't cause any harm, but if you ever want to ship something like a Debian package of Koka, it'll help if it can install everything into a logical place under `/usr/local`. Doing it in the container would be good practice since it's not possible to break anything :)

You could have something like `{prefix}/share/koka` for architecture-independent files (source or bytecode) and `{prefix}/lib/koka` for architecture-dependent ones (`.so` and other ELF files). If `{prefix}` is `/usr/local` then everything will be installed into the right places.

In practice, would it be something like this?
 
* `/usr/local/bin/koka`
* `/usr/local/lib/koka/out/...`
* `/usr/local/share/koka/samples/...`

> Otherwise very cool this works -- I will already add the Dockerfile to the repo.

Thank you!

**Comment by lassik** [comment]
Thanks for the ongoing work on the installer in the `dev` branch!

I tried the current `Dockerfile` and am getting this:

```
Step 8/17 : RUN stack exec koka -- util/install -- --prefix=/build/local
Step 9/17 : WORKDIR /build/local
Step 10/17 : RUN find /build/local | sort
/build/local
/build/local/bin
/build/local/lib
/build/local/lib/koka
/build/local/lib/koka/v2.0.6
/build/local/share
/build/local/share/koka
/build/local/share/koka/v2.0.6
```

I.e. for some reason, the build tree ends up populated only with the directory structure, not any files.

**Comment by daanx** [comment]
Strange -- I just tried myself and got errors; but then I changed the "clone" command to pull from the latest master:
```
> RUN git clone --recursive https://github.com/koka-lang/koka -b master
```
and then it did a full checkout and rebuild and now it worked fine :-)  I think docker somehow caches too much?  
The image is 435MiB now. Now, it does a standard install with 
```
/usr/local/bin/koka
/usr/local/lib/koka/v2.0.x/gcc-debug/<pre-compiled-libs>
/usr/local/lib/koka/v2.0.x/gcc-drelease/<pre-compiled-libs>
/usr/local/lib/koka/v2.0.x/gcc-release/<pre-compiled-libs>
/urs/local/share/koka/v2.0.x/lib/std/<std libs>
/urs/local/share/koka/v2.0.x/lib/samples/<std samples>
```
so all standard samples are there as well (and accessible from the interactive prompt)

So, now I just need to figure out how to upload the image to dockerhub in `kokalang` and then it should be super easy for people to run it right?

**Comment by daanx** [comment]
Ha, I think I pushed  an image that works, can you try it?
```
> docker run -it kokalang/koka:v2.0.7
```

**Comment by lassik** [comment]
Sorry for the delay. I tried it. The REPL works, as does the caesar sample if I give "bash" for the command and compile/run it from there.

To edit files, the easiest thing is to use Docker's volume mounting feature. It means you can poke a hole in the container so that a given directory inside the container will read and write a given directory on the host file system. E.g. if you start the container with `docker run -v $PWD:/foo -it kokalang/koka:v2.0.7 bash` and try `ls` inside the container, it will show the files on the host. Likewise, if you do `touch /foo/hello` in the container, the file will show up on the host as well.

It's also possible to install one or more text editors into the container, but then the question is, which ones? :)

**Comment by lassik** [comment]
I'm not managing to build the container from the current `Dockerfile`. It gives this at the end:

```
[100%] Built target kklib
Killed
The command '/bin/sh -c stack exec koka -- util/bundle -- --postfix=docker' returned a non-zero code: 137
```

Maybe means some kind of memory limit has been exceeded?

**Comment by lassik** [comment]
The usual name for `--postfix` in this context is `--suffix`.

**Comment by daanx** [comment]
Thanks for the help -- I'll try to rebuild the docker too and see if I get the same error.

- it seems the volume mounting is easiest for using source files -- once it works good I will try to doument it.
- editor wise, VS Code would be great but it may lead to a large image I guess? no idea. If you can try that would be great. 
   Otherwise we should install VIM I guess but there is currently no editor support for VIM.

**Comment by daanx** [comment]
I just updated the docker file in the latest `dev` and pushed to docker.io
```
> docker run -it kokalang/koka:v2.0.14
```
should work :-)
I now use Debian buster since Koka now uses Ghc 8.8.4 which in turn requires glibc 2.27. 

Please give it a try :-) I will probably write some docs on it soon as it seems useful -- for example to create an online "playground"/repl?

**Comment by TimWhiting** [comment]
I'm going to close this since the initial issue looks solved. Please reopen, or create a new issue if there is anything missing.
